moon/controller_round3.py

```diff
@@ -447,38 +410,81 @@ def on_scan(self, timestamp, data):
-                print ("current distance: %f, min distance so far: %f" % (straight_ahead_dist , self.homebase_final_approach_distance))
-                if straight_ahead_dist > self.homebase_final_approach_distance:
-                    if self.homebase_increase_count < MAX_NR_OF_FARTHER_SCANS:
-                        self.homebase_increase_count += 1
-                    else:
-                        self.homebase_increase_count = 0
-                        # missed it, back to main driving loop in order to try again
-                        print (self.time, "No longer tracking %s, distance increased" % self.currently_following_object['object_type'])
-                        self.currently_following_object['timestamp'] = None
-                        self.currently_following_object['object_type'] = None
-                        self.homebase_final_approach_distance = float("inf")
-                        self.basemarker_centered = False
-                        self.on_driving_control(timestamp, None) # do this last as it possibly raises exception
-                else:
-                    self.homebase_increase_count = 0
-                    self.homebase_final_approach_distance = straight_ahead_dist
+            if self.currently_following_object['object_type'] == 'homebase':
+#                print (self.time, "controller_round3: homebase current distance: %f" % (straight_ahead_dist))
 
-                    if straight_ahead_dist < HOMEBASE_KEEP_DISTANCE:
-                        self.publish("desired_movement", [0, 0, 0])
-
-                        print ("app: final homebase distance %f: " % straight_ahead_dist)
-                        self.homebase_final_approach_distance = float("inf")
-                        self.object_reached('homebase')
-                    else:
-                        # keep going straight; for now this means do nothing, keep speed from previous step
-                        self.currently_following_object['timestamp'] = self.time # freshen up timer as we are still following homebase
+                if straight_ahead_dist < HOMEBASE_KEEP_DISTANCE:
+                    self.publish("desired_movement", [0, 0, 0])
 
-            # we expect that lidar bounces off of homebase as if it was a big cylinder, not taking into consideration legs, etc.
+                    print ("app: final homebase distance %f: " % straight_ahead_dist)
+                    self.object_reached('homebase')
 
             if 'basemarker' in self.objects_to_follow:
 
+                # find min_index where distance < 5 and max_index where distance < 5
+                # find circle passing through these points (defined in polar coordinates as [index*6 degrees, distance]
+                min_index = max_index = -1
+                for i in range(0,180):
+                    if data[i]> 10 and data[i] < 5000:
+                        min_index = i
+                        break
+                for i in range(180, 0, -1):
```

Comment:
optionally you may use `numpy.array` and "mask" for faster solution if needed later ...

Reply:
as discussed on chat - this optimization can come later ..